src/cluster/ShardingConfigurationSource.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace PHPLRPM\Cluster;
+
+use PHPLRPM\ConfigurationSource;
+
+class ShardingConfigurationSource implements ConfigurationSource
+{
+    private $confSource;
+    private $clusterConf;
+
+    public function __construct(ConfigurationSource $confSource, ClusterConfiguration $clusterConf)
+    {
+        $this->confSource = $confSource;
+        $this->clusterConf = $clusterConf;
+    }
+
+    public function loadConfiguration(): array
+    {
+        $inputConfig = $this->confSource->loadConfiguration();
+        $this->clusterConf->reload();
+        $numInstances = $this->clusterConf->getNumberOfInstances();
+        $instance = $this->clusterConf->getInstanceNumber();
+        if (!self::isValidClusterConfig($numInstances, $instance)) {
+            $errorMsg = 'Invalid cluster config: ' .
+                'number of instances = ' . strval($numInstances) .
+                ', instance number = ' . strval($instance);
+            throw new ClusterConfigurationValidationException($errorMsg);
+        }
+        return self::filterConfig($inputConfig, $numInstances, $instance);
+    }
+
+    private static function isValidClusterConfig(int $numInstances, int $instance)
+    {
+        return is_int($numInstances)
+            && $numInstances > 0
+            && is_int($instance)
+            && $instance >= 0
+            && $instance < $numInstances;
+    }
+
+    private static function filterConfig(array $inputConfig, int $numInstances, int $instance): array
+    {
+        $outputConfig = [];
+        foreach ($inputConfig as $key => $value) {
+            if (crc32($key) % $numInstances == $instance) {
+                $outputConfig[$key] = $value;
+            }
+        }
```

Comment:
Since all shards are working off the same input config (which is an ordered map), you don't need to hash here. That has the benefit of evenly distributing the config entries across the shards.

```php
$index = 0;
foreach ($inputConfig as $key => $value) {
  if ($index++ % $numInstances == $instance) {
    $outputConfig[$key] = $value;
  }
}
```

Reply:
Alternatively, if it matters, you could use a consistent hash (e.g. [Rendezvous hasing](https://en.wikipedia.org/wiki/Rendezvous_hashing)) to maintain some stability as to which config entries are handled by which shard as the cluster size changes.

Reply:
I hear you, and did consider such a simple bucketing approach. 

However, such an approach would also implicitly impose the restriction that configuration arrays returned by the  same `ConfigurationSource`  on different instances would need to be not only structurally equal but also built in the same order.

While it seems highly likely that any possible `ConfigurationSource` will always insert jobs in the same order when running on different instances, perhaps it would be good to have a stronger guarantee. I'm imagining a case where these keys (ids) might come from some external source that might not return them in the same predictable order for every client request.

Reply:
> maintain some stability as to which config entries are handled by which shard as the cluster size changes

This I also thought about and it's a feature I can imagine as being nice to have for some use cases! In use cases where scaling up and down does not involve restarting the world, minimal service disruption on up/down-scaling could be a desirable property. lrpm will check the configuration diff and will not disturb processes that persist between configuration changes.

Reply:
Implemented both sharding strategies, moved code to separate repo:

https://github.com/vrza/php-lrpm-cluster

Rendezvous hashing lib is also in a separate repo:

https://github.com/vrza/php-rendezvous-hashing